### PR TITLE
chore(parser): Parser error optimisation

### PR DESCRIPTION
**Pull request by joss-aztec**
# Related issue(s)

Resolves # 

# Description

## Summary of changes

tl;dr
~40% speed up in release mode
~10% speed up in debug mode

Chumsky produces and aggregates large numbers of errors in normal operation regardless of whether it's on the happy path. This is because as it tries each grammar rule in turn, it produces an error until the matching rule is encountered. Currently our error construction and aggregation is rather expensive, largely in due to the construction and destruction of `BTreeSet`s that may hold zero or very few items.

This PR uses `SmallOrdSet` to avoid heap allocations as long as possible.

Other changes:
- labels switched to an enum to avoid string allocations

Samples taken using `time nargo print-acir`

|              | debug  | release |
|--------|--------|---------|
| master | 4.1s      |     0.47s |
| this branch | 3.7s |  0.34s |

## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

**Review by jfecher — COMMENTED** [review]
Looks pretty good, I appreciate the timing information as well. Have you tested against existing small set crates? There aren't many but I wonder it one of those fits our usecase any better.

File: crates/noirc_frontend/src/parser/errors/late_alloc_set.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+use std::collections::BTreeSet;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum LateAllocSetData<T> {
+    None,
+    One(T),
+    Two(T, T),
+    Three(T, T, T),
+    Set(BTreeSet<T>),
+}
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_frontend/src/parser/errors/late_alloc_set.rs:33]: I'd also prefer to use an existing crate -- I have not looked at those mentioned but I know for vectors, its smallvec.

Main benefit being that there's less code to maintain

File: crates/noirc_frontend/src/parser/labels.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+use std::fmt;
+
+use crate::token::TokenKind;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
```
Comment (kevaundray) [inline crates/noirc_frontend/src/parser/labels.rs:5]: Can you put a comment for what this is? Something like "constants used in the parser"

Not sure about the name `ParserLabel` but want to avoid bikeshedding :)

File: crates/noirc_frontend/src/parser/errors/late_alloc_set.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (joss-aztec) [reply crates/noirc_frontend/src/parser/errors/late_alloc_set.rs:33]: Just tried small-ord-set:
debug: 1700±100ms
release: 182±8ms
Slower in debug, but faster in release - let's got with that.

**Comment by joss-aztec** [comment]
Just noticed these changes throw up the following clippy warning:

```
warning: the `Err`-variant returned from this function is very large
   --> crates/noirc_frontend/src/ast/mod.rs:154:10
    |
154 |     ) -> Result<UnresolvedTypeExpression, ParserError> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the `Err`-variant is at least 344 bytes
    |
    = help: try reducing the size of `parser::errors::ParserError`, for example by boxing large elements or replacing it with `Box<parser::errors::ParserError>`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err
    = note: `#[warn(clippy::result_large_err)]` on by default
```

Guess I better benchmark `Resolver::resolve_expression` to check I haven't inadvertently hurt it. If not, then I guess it's justifiable to mute the warning.

I'm aware that I've spent more time on this PR than it's really worth - so shout if you think I should abandon.

**Comment by jfecher** [comment]
I think this PR is still a worthwhile pursuit, its nice to have the timing information so that we know for later what any performance characteristics are like. I'm interested in what the performance of tinyset would be as well? Is it an easy slot-in replacement for small-ord-set?

**Comment by joss-aztec** [comment]
I had a go at `tinyset`, but it's pretty cumbersome to integrate with our current enums. You need to implement the following trait on your element type:
```rs
pub trait Fits64: Copy {
    /// Convert back *from* a u64.  This is unsafe, since it is only
    /// infallible (and lossless) if the `u64` originally came from
    /// type `Self`.
    unsafe fn from_u64(x: u64) -> Self;
    /// Convert to a `u64`.  This should be infallible.
    fn to_u64(self) -> u64;
}
```
Probably the biggest hurdle is you'd need to map to some alternative representation of a token that's able to support the `Copy` trait. (Maybe an argument to split data from ids as we have done for ssa?) I also had trouble with `unsafe` related errors, but that could be my rust inexperience rather than a real blocker.

**Review by jfecher — DISMISSED** [review]
Looks good. I have one final question but I don't want to hold this PR any longer.

File: crates/noirc_frontend/src/parser/errors.rs [hunk 3]
```diff
@@ -1,19 +1,41 @@
...
+/// and as such, the time taken to do so significantly impacts parsing performance. For this
+/// reason we use `SmallOrdSet` to avoid heap allocations for as long as possible - this greatly
+/// inflates the size of the error, but this is justified by a resulting increase in parsing
+/// speeds of approximately 40% in release mode.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct ParserError {
-    expected_tokens: LateAllocSet<Token>,
-    expected_labels: LateAllocSet<ParserLabel>,
+    expected_tokens: SmallOrdSet<[Token; 3]>,
+    expected_labels: SmallOrdSet<[ParsingRuleLabel; 3]>,
```
[the remark this answers is not in the document]
Comment (joss-aztec) [reply crates/noirc_frontend/src/parser/errors.rs]: Originally it was because I found that 90% of errors go accumulate 3 tokens. However, I've just sampled at different sizes and found that they might as well be sized 1.

**Review by jfecher — APPROVED** [review]
